### PR TITLE
C# wrapper types

### PR DESCRIPTION
**Pull request by jskeet**
This is ready for an initial review, but I wouldn't be surprised if we wanted some more changes before merging.

@jtattermusch @anandolee

### CODE REVIEW

File: csharp/src/ProtocolBuffers.Test/WellKnownTypes/WrappersTest.cs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+                    { 10000, null }
+                },
+                DoubleField = { { 1, 12.5 }, { 10, -1.5 }, { 20, 0d } },
+                FloatField = { { 2, 123.25f }, { 3, -20f }, { 4, 0f } },
+                Int32Field = { { 5, int.MaxValue }, { 6, int.MinValue }, { 7, 0 } },
+                Int64Field = { { 8, long.MaxValue }, { 9, long.MinValue }, { 10, 0L } },
+                StringField = { { 11, "First" }, { 12, "Second" }, { 13, "" }, { 14, null } },
+                Uint32Field = { { 15, uint.MaxValue }, { 16, uint.MinValue }, { 17, 0U } },
+                Uint64Field = { { 18, ulong.MaxValue }, { 19, ulong.MinValue }, { 20, 0UL } },
+            };
```
Comment (jtattermusch) [inline csharp/src/ProtocolBuffers.Test/WellKnownTypes/WrappersTest.cs]: This test doesn't test anything.

Comment (jskeet) [inline csharp/src/ProtocolBuffers.Test/WellKnownTypes/WrappersTest.cs]: Doh! Will fix :)

**Comment by jskeet** [comment]
Done.

File: csharp/src/ProtocolBuffers.Test/WellKnownTypes/WrappersTest.cs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+            var fields = ((IReflectedMessage) message).Fields;
+            var dictionary = (IDictionary) fields[MapWellKnownTypes.Int32FieldFieldNumber].GetValue(message);
+            Assert.AreEqual(2, dictionary[1]);
+            Assert.IsNull(dictionary[3]);
+            Assert.IsTrue(dictionary.Contains(3));
+        }
+
+        // Merging is odd with wrapper types, due to the way that default values aren't emitted in
+        // the binary stream. In fact we cheat a little bit - a message with an explicitly present default
+        // value will have that default value ignored.
```
Comment (jtattermusch) [inline csharp/src/ProtocolBuffers.Test/WellKnownTypes/WrappersTest.cs]: Let's add a test that parses an IntWrapper from a stream that has the default value set explicitly. Comment of that test can describe that this is actually not the ideal behavior, but that's the way it's implemented now.

Comment (jskeet) [inline csharp/src/ProtocolBuffers.Test/WellKnownTypes/WrappersTest.cs]: Happy to do so, but will be in the morning.

File: csharp/src/ProtocolBuffers/Collections/MapField.cs [hunk 3]
```diff
@@ -51,14 +51,38 @@ namespace Google.Protobuf.Collections
     public sealed class MapField<TKey, TValue> : IDeepCloneable<MapField<TKey, TValue>>, IFreezable, IDictionary<TKey, TValue>, IEquatable<MapField<TKey, TValue>>, IDictionary
     {
         // TODO: Don't create the map/list until we have an entry. (Assume many maps will be empty.)
+        private bool allowNullValues;
```
Comment (jtattermusch) [inline csharp/src/ProtocolBuffers/Collections/MapField.cs]: readonly

File: csharp/src/ProtocolBuffers/FieldCodec.cs [hunk 4]
```diff
@@ -131,6 +131,106 @@ public static FieldCodec<T> ForEnum<T>(uint tag, Func<T, int> toInt32, Func<int,
...
+            var nestedCodec = WrapperCodecs.GetCodec<T>();
+            return new FieldCodec<T?>(
+                input => WrapperCodecs.Read<T>(input, nestedCodec),
+                (output, value) => WrapperCodecs.Write<T>(output, value.Value, nestedCodec),
+                value => value == null ? 0 : WrapperCodecs.CalculateSize<T>(value.Value, nestedCodec),
+                tag,
+                null); // Default value for the wrapper
+        }
+
+        // Helper code to create codecs for wrapper types. Somewhat ugly with all the
```
Comment (jtattermusch) [inline csharp/src/ProtocolBuffers/FieldCodec.cs]: missing end of sentence?

File: csharp/src/ProtocolBuffers/FieldCodec.cs [hunk 5]
```diff
@@ -131,6 +131,106 @@ public static FieldCodec<T> ForEnum<T>(uint tag, Func<T, int> toInt32, Func<int,
...
+                tag,
+                null); // Default value for the wrapper
+        }
+
+        // Helper code to create codecs for wrapper types. Somewhat ugly with all the 
+        private static class WrapperCodecs
+        {
+            private static readonly Dictionary<Type, object> Codecs = new Dictionary<Type, object>
+            {
+                { typeof(bool), ForBool(WireFormat.MakeTag(1, WireFormat.WireType.Varint)) },
```
Comment (jtattermusch) [inline csharp/src/ProtocolBuffers/FieldCodec.cs]: I'd extract a constant for the field number.
// All wrappers keep the value in field with number 1.
const int WrapperValueFieldNumber = 1

File: src/google/protobuf/compiler/csharp/csharp_wrapper_field.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  GOOGLE_DISALLOW_EVIL_CONSTRUCTORS(WrapperOneofFieldGenerator);
+};
+
+}  // namespace csharp
+}  // namespace compiler
+}  // namespace protobuf
+}  // namespace google
+
+#endif  // GOOGLE_PROTOBUF_COMPILER_CSHARP_WRAPPER_FIELD_H__
+
```
Comment (jtattermusch) [inline src/google/protobuf/compiler/csharp/csharp_wrapper_field.h]: extra line at EOF

File: src/google/protobuf/unittest_well_known_types.proto [hunk 7]
```diff
@@ -27,5 +30,83 @@ message TestWellKnownTypes {
...
+  google.protobuf.Int64Value int64_field = 12;
+  google.protobuf.UInt64Value uint64_field = 13;
+  google.protobuf.Int32Value int32_field = 14;
+  google.protobuf.UInt32Value uint32_field = 15;
+  google.protobuf.BoolValue bool_field = 16;
+  google.protobuf.StringValue string_field = 17;
+  google.protobuf.BytesValue bytes_field = 18;
+}
+
+// A repeated field for each well-known type.
```
Comment (jtattermusch) [inline src/google/protobuf/unittest_well_known_types.proto]: Just a note: we should probably point out this change once merging to the master.

**Comment by jtattermusch** [comment]
Thanks for contributing this!
The changes LGTM, but I had some minor comments.
Because of timezone issues, feel free to merge this PR yourself once you address them.

File: csharp/src/ProtocolBuffers/FieldCodec.cs (hunk 5 above)
Comment (jskeet) [inline csharp/src/ProtocolBuffers/FieldCodec.cs]: Better:

private const int WrapperValueFieldNumber = Google.Protobuf.WellKnownTypes.Int32Value.ValueFieldNumber;

As we're generating it, we might as well use it :)

(But I've still got the constant - I'm not using each individual wrapper.)

File: csharp/src/ProtocolBuffers/FieldCodec.cs (hunk 4 above)
Comment (jskeet) [inline csharp/src/ProtocolBuffers/FieldCodec.cs]: Done (and made it XML documentation)

File: csharp/src/ProtocolBuffers/Collections/MapField.cs (hunk 3 above)
Comment (jskeet) [inline csharp/src/ProtocolBuffers/Collections/MapField.cs]: Thanks, fixed.

File: src/google/protobuf/compiler/csharp/csharp_wrapper_field.h (hunk 6 above)
Comment (jskeet) [inline src/google/protobuf/compiler/csharp/csharp_wrapper_field.h]: Done.

File: csharp/src/ProtocolBuffers.Test/WellKnownTypes/WrappersTest.cs (hunk 2 above)
Comment (jskeet) [inline csharp/src/ProtocolBuffers.Test/WellKnownTypes/WrappersTest.cs]: Done - and added a fairly simple oneof test too.

**Comment by jskeet** [comment]
Thanks. Pushed the changes - waiting for CI, then will merge.

**Pull request merged**